spacy/pipeline/span_finder.py

```diff
@@ -0,0 +1,395 @@
+    training_key: str = DEFAULT_SPANS_KEY,
+    **kwargs,
+) -> Dict[str, Any]:
+    kwargs = dict(kwargs)
+    attr_prefix = "span_finder_"
+    kwargs.setdefault("attr", f"{attr_prefix}{predicted_key}")
+    kwargs.setdefault("allow_overlap", True)
+    kwargs.setdefault(
+        "getter", lambda doc, key: doc.spans.get(key[len(attr_prefix) :], [])
+    )
+    kwargs.setdefault("labeled", False)
+    kwargs.setdefault("has_annotation", lambda doc: predicted_key in doc.spans)
+    # score_spans can only score spans with the same key in both the reference
+    # and predicted docs, so temporarily copy the reference spans from the
+    # reference key to the candidates key in the reference docs, restoring the
+    # original span groups afterwards
+    orig_span_groups = []
+    for eg in examples:
+        orig_span_groups.append(eg.reference.spans.get(predicted_key))
+        if training_key in eg.reference.spans:
+            eg.reference.spans[predicted_key] = eg.reference.spans[training_key]
+    scores = Scorer.score_spans(examples, **kwargs)
+    for orig_span_group, eg in zip(orig_span_groups, examples):
+        if orig_span_group is not None:
+            eg.reference.spans[predicted_key] = orig_span_group
+    return scores
+
+
+class _MaxInt(int):
+    def __le__(self, other):
+        return False
+
+    def __lt__(self, other):
+        return False
+
+    def __ge__(self, other):
+        return True
+
+    def __gt__(self, other):
+        return True
+
+
+class SpanFinder(TrainablePipe):
+    """Pipeline that learns span boundaries"""
+
+    def __init__(
+        self,
+        nlp: Language,
+        model: Model[Iterable[Doc], Floats2d],
+        name: str = "span_finder",
+        *,
+        threshold: float = 0.5,
+        max_length: Optional[int] = None,
+        min_length: Optional[int] = None,
+        # XXX I think this is weird and should be just None like in
+        scorer: Optional[Callable] = partial(
+            span_finder_score,
+            predicted_key=DEFAULT_PREDICTED_KEY,
+            training_key=DEFAULT_SPANS_KEY,
+        ),
+        predicted_key: str = DEFAULT_PREDICTED_KEY,
+        training_key: str = DEFAULT_SPANS_KEY,
+    ) -> None:
+        """Initialize the span boundary detector.
+        model (thinc.api.Model): The Thinc Model powering the pipeline component.
+        name (str): The component instance name, used to add entries to the
+            losses during training.
+        threshold (float): Minimum probability to consider a prediction
+            positive.
+        scorer (Optional[Callable]): The scoring method.
+        predicted_key (str): Name of the span group the candidate spans are saved to
+        training_key (str): Name of the span group the training spans are read from
+        max_length (int): Max length of the produced spans (unlimited when set to 0)
+        min_length (int): Min length of the produced spans (unlimited when set to 0)
```

Comment:
Update the types and the descriptions here, too. I'd add `Defaults to None` for users who are just reading the docstrings.